Paper: CLAIMED, a visual and scalable component library for Trusted AI

fatkaratekid opened the pull request:
I am submitting a full paper, as mentioned in the email titled: "Invitation to submit a full paper for SciPy 2021"

The paper builds successfully on my machine.

-------

If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@deniederhut<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@stsievert, @cmarmo<!--end-reviewers-list-->

fatkaratekid committed:
add abstract

fatkaratekid committed:
add title and keywords

fatkaratekid committed:
add authors

fatkaratekid committed:
add introduction

fatkaratekid committed:
add kubernetes section

fatkaratekid committed:
add deep learning with tensorflow section

fatkaratekid committed:
add kubeflow section

fatkaratekid committed:
add elyraai section

fatkaratekid committed:
add jupyterlab section

fatkaratekid committed:
add ai explainability section

fatkaratekid committed:
add AI Fairness and Bias section

fatkaratekid committed:
add AI Adversarial Robustness section

fatkaratekid committed:
add pipelines, input, transform, filter components

fatkaratekid committed:
fix formating

fatkaratekid committed:
add components image transformer, training, evaluation, blessing, publishing sections

fatkaratekid committed:
add future work section

fatkaratekid committed:
add conclusion section

fatkaratekid committed:
add incomplete references

fatkaratekid committed:
fix bias reference link

fatkaratekid committed:
fix references

fatkaratekid committed:
fix wiki and art references

fatkaratekid committed:
fix references

fatkaratekid committed:
fix formatting

fatkaratekid committed:
fix references and errors

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

deniederhut commented:
@scoobies assign @deniederhut as editor

scoobies commented:
Assigned! @deniederhut is now the editor

deniederhut commented:
@scoobies add @stsievert to reviewers

scoobies commented:
@stsievert added to the reviewers list!

deniederhut commented:
We've been having difficulty finding reviewers for some papers, so to make sure everyone has enough time to do their reviews, we will be bumping the deadline for the initial review back to June 17th.

deniederhut commented:
@scoobies add @cmarmo to reviewers

scoobies commented:
@cmarmo added to the reviewers list!

deniederhut commented:
Hi @stsievert and @cmarmo ! Friendly reminder that the initial review deadline is next week

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+balance the load etc. Although there are other options like Docker
+Swarm, Kubernetes is the market leader in this domain. It was donated to
+CNCF [cncf]_ by Google, which means a lot of Google’s know-how and years of
+experience went into it. The system can run on public, on-prem or on
```
```suggestion
experience went into it. The system can run on public, on-premises or on
```
I am in favor of avoiding abridged words when possible (see also below), for clarity.

fatkaratekid replied:
This really does make sense, and I am a proponent of it myself. The only issue with the proposed change is that the industry standard for the term is **on-premise**, which rubs some grammar people the wrong way. In that camp, the consensus is we should use **on-premises**. Finally, most people from both camps tend to at least tolerate **on-prem**.
Adam Fowler had some thoughts on it too ([link](https://www.adamfowlerit.com/2017/04/premise-vs-premises-cares/)) and, based on the last update from a linguist in the comment section, it's even alright for us to use the term **on-premise**. It's a conundrum.

The other one is clear (the one below).

cmarmo replied:
Even without being an expert, when I read 'on-premises' or 'on-premise' I can guess the meaning if I know a bit of English, because I have a full word to understand. With 'on-prem' I need to search for it on internet, because this is an abbreviation... that's what I did and I learned something, thanks for that...
To quote your reference "On-Premise vs On-Premises – Who Cares?", both are closer to plain English than 'on-prem', in my opinion.

fatkaratekid replied:
thank you, changed to on-premises

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+communication between containers, schedule them, manage authentications,
+balance the load etc. Although there are other options like Docker
+Swarm, Kubernetes is the market leader in this domain. It was donated to
+CNCF [cncf]_ by Google, which means a lot of Google’s know-how and years of
```
I believe this is the first occurrence of the CNCF acronym: do you mind giving the extended definition here? Thanks.

fatkaratekid replied:
Added full definition in this commit 2c8b80e

fatkaratekid replied:
thanks for pointing it out

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+CNCF [cncf]_ by Google, which means a lot of Google’s know-how and years of
+experience went into it. The system can run on public, on-prem or on
+hybrid clouds. On-prem installation is very important for institutions
+dealing with sensitive data. For IBM, Kubernetes is also strategic,
+joining CNCF, having moved all Watson Services to Kubernetes and aquired
+RedHat, IBM is now 3rd largest committer to Kubernetes [ibmcncf]_.
```
I find the sentence too compact: Kubernetes is strategic to IBM because it is the 3rd largest committer, or because it moved all Watson services to Kubernetes? Or it became largest committer because they moved to Kubernetes... or ...
Do you mind expanding the sentence? This will clarify the cause - effect relations. Thanks.

romeokienzler replied:
Thanks! Let me put the following in if you are ok: 
For IBM, Kubernetes is also strategic. This is mainly because Kubernetes enables the hybrid cloud scenario backed by an
open source common runtime capable of transparently moving workload across different on-prem, remote and cloud data centers seamlessly. Besides acting as a Kubernetes runtime provider in the cloud and - through the aquisition of RedHat - becoming the major vendor for on-prem Kubernetes, IBM is now able to deliver software solution on top of Kubernetes, making them run everywhere. Therefore, IBM joined CNCF,  moved all Watson Services to Kubernetes and aquired RedHat. This makes IBM the 3rd largest committer to Kubernetes.

romeokienzler replied:
added to document in commit b76ef663e9fc7edffb5fa2b0b2f11a1632cf9c0b

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+Deep Learning with TensorFlow
+-----------------------------
+
+It is the second incarnation of the Google Brain project’s scalable
```
```suggestion
TensorFlow is the second incarnation of the Google Brain project’s scalable
```

fatkaratekid replied:
much clearer, thanks

stsievert commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+
+   Deep Learning models are getting more and more popular but constraints on explainability, adversarial robustness and fairness are often major concerns for production deployment. Although the open source ecosystem is abundant on addressing those concerns, fully integrated, end to end systems are lacking in open source.
+
+   Therefore we provide an entirely open source, reusable component framework, visual editor and execution engine for production grade machine learning on top of Kubernetes, a joint effort between IBM and the University Hospital Basel. It uses Kubeflow Pipelines, the AI Explainability360 toolkit, the AI Fairness360 toolkit and the Adversarial Robustness Toolkit on top of ElyraAI, Kubeflow, Kubern ...
```
> We explain how we’ve created a COVID-19 deep learning classification pipeline based on CT scans. We use the toolkit to highlight parts of the images which have been crucial for the model’s decisions. We detect bias against age and gender and finally, show how to deploy the model to KFServing to share it across different hospital data centers of the Swiss Personalized Health Network.

That's a lot of words to say "To exemplify the use of CLAIMED, we show a toy workflow to train, debug and deploy of a deep learning model that determines if CT scans are positive for COVID-19."

When I first read this abstract, I expected to see detailed experiments on age and gender, detailed experiments on COVID-19 image classification (that had reasonable accuracy), and detailed experiments on deployment (including config scripts).

stsievert replied:
> an entirely open source, reusable component framework, visual editor and execution engine

What does your software do, or what actions are possible with your software? I'm not sure why I'd care about a "visual editor," "reusable component framework" or "execution engine."

Why do you mention Jupyter notebooks, ElyraAI, KubeFlow, Kubernetes and JupyterLab in the abstract? What *features* do they add to your tool?

fatkaratekid replied:
This was addressed as of this commit 9cb2543.

stsievert commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+component library and figure :ref:`kfp` is a screenshot taken from
+Kubeflow displaying the pipeline after finishing it’s run.
+
+.. figure:: elyra_pipeline.png
```
This figure is not readable. I can't determine the function of any of the blocks. It might be more clear to have a simpler and readable workflow (as expressed on *[AI Pipelines](https://elyra.readthedocs.io/en/stable/user_guide/pipelines.html)*).

fatkaratekid replied:
Just to understand correctly, you suggest the image should contain explanations like on the Elyra page you provided?
e.g. like this
![image](https://user-images.githubusercontent.com/2835130/122767013-77b7c800-d2a2-11eb-860f-4a20e35b5655.png)

romeokienzler replied:
I think it is too small. I've zoomed in and stripped away unnecessary parts - can you please have a [look](https://github.com/fatkaratekid/scipy_proceedings/commit/cb1332a6a970b8bdbd0f449cd53c08ca40c4ae15) now?

romeokienzler replied:
commit cb1332a6a970b8bdbd0f449cd53c08ca40c4ae15

stsievert commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+
+    Kubernetes, Kubeflow, JupyterLab, ElyraAI, KFServing, TrustedAI, AI Explainability, AI Fairness, AI Adversarial Robustness
+
+Introduction
```
I think good introductions give solid reasoning for the software tool. What problem would you like to address, and why is your software necessary? What users would find it useful?

I find an answer to that question a lot more valuable in the introduction then "what is our software built from?"

stsievert replied:
The subsections in the intro are below:

* Containerization and Kubernetes
* Deep Learning with TensorFlow
* Kubeflow
* ElyraAI
* JupyterLab
* AI Explainability
* AI Fairness and Bias
* AI Adversarial Robustness

Which sections describe core features of your library? Which sections describe the implementation of those core features?

fatkaratekid replied:
The introduction has been modified to better answer suggested questions and make it more readable (commit 65f7ac0a).

stsievert commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+categories paired with a description of exemplary components used in
+health care. This pipeline is also available in open source  [ect]_.
+
+Containerization and Kubernetes
```
Why are these sections in the introduction? Do they introduce your software and explain why your software is necessary?

Generally, I care a lot more about what your software can be used for and what features it has. After I understand the problem, I'll look into how the software is built.

fatkaratekid replied:
This is fixed in 9cb25432. I hope it is good now, but additional input is always welcome.

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+abundant, but the community lacks a fully integrated, trusted and
+scalable visual tool. Therefore we have built CLAIMED, the visual
+**C**\ omponent **L**\ ibrary for **AI**, **M**\ achine Learning,
+**E**\ TL and **D**\ ata Science which runs on top of ElyraAI capable of
```
I am a bit uncomfortable with this acronym ETL used to define the main acronym CLAIMED: I understand ETL is meant to be a very known acronym, perhaps a footnote expanding ETL could be added? IBM itself propose a nice [explanatory page](https://www.ibm.com/cloud/learn/etl) about ETL.

stsievert replied:
I'd appreciate a better definition of ETL; I'm unfamiliar with that acronym.

fatkaratekid replied:
How does this look to you 6021b270?

stsievert commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+
+   The exemplary TrustedAI pipeline for the health care use case. :label:`pipeline`
+
+Pipeline Components
```
Is this section it explanation of a particular set of components? Or is a list of every set of components? I'd appreciate both; isn't the pipeline the core feature of your software?

romeokienzler replied:
This section exemplifies each (currently existing) category with at least one component which has been used for this particular pipelines. There exist more components not part of this pipeline. The core feature of our software is threefold.

- the CLAIMED component library
- Elyra (which is developed in open source by the IBM team I'm working with) with it's capability to use CLAIMED to create a pipeline and push it to KubeFlow
- the pipeline itself

please see respective commit 23cd1d7233aa8987163b14d09a72ae03837126bd

stsievert commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+Besides their stunning performance, deep learning models face a lot of
+resistance for production usage because they are considered to be a
+black box. Technically (and mathematically) deep learning models are a
+series of non-linear feature space transformations - sounds scary, but
```
Style nit: I'd remove this sentence. A "non-linear feature space transformation" does not "sound scary" to me.

fatkaratekid replied:
This should be good now, fixed in 4c0858f, or check the newer commit 35ca321.

stsievert commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+   Example on how LIME helps to identify classification relevant
+   areas of an image. :label:`limefig`
+
+AI Fairness and Bias
```
> It turns out that it isn’t that simple.
> Fairness assessment and bias detection is an art on it’s own.

Citations needed.

romeokienzler replied:
removed artistic expressions in commit 7f77346902d39ff3e37e0b71fe7ba495b972d0d3

please let me know if this is what you've imagined

stsievert reviewed (commented):
Thanks for the submission @fatkaratekid! This looks like an interesting tool to manage machine learning (ML) pipelines. IIUC, default pipeline components help with model training, interpretation, fairness, and robustness (through Tensorflow, LIME, AI360 and ART respectively). CLAIMED pipelines are created with JupyterLab and can be executed on a Kubernetes cluster/Docker Swarm.

Some questions I have are below:

1. What's required to run this software locally?
2. Why doesn't the abstract only mention the problem CLAIMED solves and the features CLAIMED provides? Someone who picks up this paper will care most about the features of the software, not the underlying building blocks. Generally, it took a while to understand this paper because I got distracted by the underlying building blocks (Kubernetes, Tensorflow, etc).
3. I'm still not certain about what this software is capable of. What pipeline components are included by default? Can I develop a custom component? e.g., what if I want to use PyTorch instead of Tensorflow?

I also have some more context-specific comments:

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+also being used in demanding production environments. On a development
+side, representing machine learning algorithms in a tree-like structures
+makes it a very good expression interface. Lastly, on the performance vs
+usability side, both eager and graph modes are supported. Meaning
```
If I understand correctly eager execution is specific to TensorFlow, for non expert readers it could be useful to have a note explaining what eager execution is (the [TensorFlow documentation](https://www.tensorflow.org/guide/eager) is quite clear itself).

fatkaratekid replied:
I hope this explains it better da6bc79c

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+
+Besides their stunning performance, deep learning models face a lot of
+resistance for production usage because they are considered to be a
+black box. Technically (and mathematically) deep learning models are a
```
Feel free to choose between ''Technically" and "mathematically", the two of them are redundant.

fatkaratekid replied:
Fixed in  35ca321

fatkaratekid replied:
Thanks for pointing it out!

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+series of non-linear feature space transformations - sounds scary, but
+in other words, per definition it is very hard to understand the
+individual processing steps a deep learning network performs. But
+techniques exist to look over a deep learning model’s shoulder. The one
```
This is a suggestion to avoid unnecessary words and subjective claims.
```suggestion
series of non-linear feature space transformations, it is hard to understand the
individual processing steps a deep learning network performs.
Techniques exist to look over a deep learning model’s shoulder. The one
```

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+
+The de facto standard for labeled image data is putting images into one
+folder per class/category. But in this particular case, the raw data
+isn’t in the required format. It’s just a folder full of images and
```
There is no need to refer to a standard that is not followed by the actual data, especially if this standard is not documented.
A simple description of the data structure will be enough and will improve readability
```suggestion
The raw images are contained in a folder and
```

fatkaratekid replied:
Thanks for taking the time to make these suggestions. I have rewritten the whole part to make it more readable. Please take a look at this commit e0ccca0

I hope it is now clear that we want to convert the data into the structure that can be consumed by TF's API.

fatkaratekid replied:
A small modification to the text also done by the commit 6ccb03b

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+
+.. figure:: images_folder_tree.png
+
+   De facto standard in folder structure for image classification data. :label:`imgdir`
```
As per comment ~below~ above
```suggestion
   Final folder structure for image classification data. :label:`imgdir`
```

fatkaratekid replied:
This has been resolved by the same commit as the one above e0ccca0

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,360 @@
+Training Components
+~~~~~~~~~~~~~~~~~~~
+
+Understanding, defining and training deep learning models is an art on
+it’s own. Training a deep learning image classification model requires a
```
> Understanding, defining and training deep learning models is an art on its own. Luckily ...

This sentence appears twice in the text, literally. In both cases, is it really necessary?

romeokienzler replied:
addressed by 7f77346902d39ff3e37e0b71fe7ba495b972d0d3

cmarmo reviewed (commented):
Dear @fatkaratekid, thanks for your submission.
As stressed in your paper, automation in Machine Learning as in any other data processing task is a guarantee of reproducibility.
Having a background in astronomy and scientific data processing I particularly value the effort of building standardized pipelines using open tools.
While I am not an expert in Machine Learning, I have some general comments about this proceeding paper.
1) Sections from "Containerization and Kubernetes" to "AI Adversarial Robustness" don't belong to the Introduction. Their titles (then the subsection contents) are not consistent: sometimes the tool is presented together with the issue it is meant to solve (eg "Containerization and Kubernetes"), sometimes only the tool ("JupyterLab"), sometimes only the issue to be addressed ("AI Adversarial Robustness"). It is important to cite in this context which tool is used, but in my understanding this is more a section describing general pipeline steps. Giving more consistent structure to the subsections will help the reader in understanding how your application can be reused.
2) I suggest to avoid 'useless words' and sentences like 'just', 'really', 'very', ... they add no information to the text and do not feel objectives, in the majority of cases. I have deleted some in my inline suggestions. The text will gain in clarity.

Commit:
adjusted abstract

deniederhut commented:
@scoobies mark pending comment

fatkaratekid committed:
correct term

Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>

fatkaratekid committed:
reduce ambiguity

Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>

Commit:
improved introduction

Commit:
merged conflict

Commit:
reorganize introduction

Commit:
create requirements mapping table against execution engines

fatkaratekid committed:
modify table O size to match X

fatkaratekid committed:
remove unnecessary words

Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>

fatkaratekid committed:
remove redundant word

Commit:
add requirements description for execution engine

fatkaratekid committed:
add full name for CNCF

Commit:
add non-working example for table caption

Commit:
Merge branch '2021' of github.com:fatkaratekid/scipy_proceedings into 2021

Commit:
fix non-working example for table caption

romeokienzler commented:
Meghann Agarwal (@mepa)
    Chris Calloway (@cbcunc)
    Dillon Niederhut (@deniederhut)
    David Shupe (@stargaser)

Do you have an example on a table in rst which successfully compiles in latex, the following doesn't work, if you remove the table caption it works:

.. table:: Requirements for execution engines. KF: Kubeflow, AF: Airflow, SM: Snakemake
================== == == ===== == ==== ======== =====
Requirement        KF AF Slurm SM Qsub HTCondor Reana
================== == == ===== == ==== ======== =====
Kubernetes Support X  X  O     X  O    X        X
GPU support        X  X  X     X  X    X        X
Component Library  X  O  O     O  O    O        O
Reproducibility    X  X  O     X  O    X        X
Data Lineage       X  O  O     O  O    O        X
================== == == ===== == ==== ======== =====

it gives me:

(/home/romeokienzler/gitco/scipy_proceedings/texlive/texlive/2021/texmf-dist/te
x/latex/psnfss/t1pcr.fd)
! Argument of \@xverbatim has an extra }.
<inserted text> 
                \par 
l.487 backrefs: }
                 
!  ==> Fatal error occurred, no output PDF file produced!
Transcript written on paper.log.

You can reproduce by  
git checkout 33f0b906609d10c38958351b6c5a3a7b5cec51a4 
on https://github.com/fatkaratekid/scipy_proceedings

fatkaratekid committed:
add dataset reference

fatkaratekid committed:
edit CLAIMED name explanation

fatkaratekid committed:
add tf eager explanation

fatkaratekid commented:
@romeokienzler 
Based on the examples from the repo, it seems it should look like this:
```
.. table:: This is the caption for the materials table. :label:`mtable`

   +------------+----------------+
   | Material   | Units          |
   +============+================+
   | Stone      | 3              |
   +------------+----------------+
   | Water      | 12             |
   +------------+----------------+
   | Cement     | :math:`\alpha` |
   +------------+----------------+
```

or this for a wide table
```
.. table:: This is the caption for the wide table.
   :class: w

   +--------+----+------+------+------+------+--------+
   | This   | is |  a   | very | very | wide | table  |
   +--------+----+------+------+------+------+--------+
```

I am not sure if there is a better way.

Commit:
add requirments and table for integrated tools

Commit:
Merge branch '2021' of github.com:fatkaratekid/scipy_proceedings into 2021

Commit:
add Final technology choice

Commit:
add diagrams presentation

Commit:
add architecture diagram

Commit:
add transition to UC

romeokienzler commented:
> `` .. table:: This is the caption for the materials table. :label:`mtable` ``

this also breaks the build

Commit:
zoom into elyra pipeline

Commit:
add zoomed elyra pipeline image

fatkaratekid committed:
rewrite Image Transformer Components

fatkaratekid committed:
fix typos

fatkaratekid committed:
modify Image Transformer Components section

Commit:
improve architecture diagram

Commit:
add further explainations on the nature of the solution

Commit:
remove artistic expression

Commit:
explain IBM Kubernetes stategy

fatkaratekid commented:
@deniederhut, @cmarmo and @stsievert, thank you all for your feedback and your contribution. It really helped us a lot to turn this into a better paper. While we give you time to make sure we have addressed all of the points you have raised, we will go through the paper again and make sure that it flows correctly. After the review period is over or as soon as we have addressed all the issues, we will send the paper for the proofreading and make those final adjustments.

fatkaratekid committed:
remove word redundancy in the abstract

fatkaratekid committed:
fix typo

fatkaratekid committed:
make small modifications throughout the document

Commit:
add future work section

Commit:
proofread first part

fatkaratekid committed:
modify future work section

Commit:
finalize quality check

deniederhut commented:
@fatkaratekid to give everyone more time to respond to their reviews, we've pushed back the deadline for authors' final revisions to July 1st

fatkaratekid committed:
make changes to future work

fatkaratekid committed:
modify future work and conclusion

fatkaratekid committed:
fix table and the caption

romeokienzler commented:
@deniederhut, @cmarmo and @stsievert I think we are in good shape now - would you be able to have a 2nd look at the document please? @fatkaratekid

fatkaratekid committed:
fix build breaking

fatkaratekid committed:
fix both tables and references

fatkaratekid committed:
fix KubeFlow

fatkaratekid committed:
rephrase different parts

fatkaratekid committed:
change table captions

Commit:
revised conclusion

fatkaratekid committed:
adjust conclusion

fatkaratekid committed:
modify abstract

fatkaratekid committed:
adjust abstract

fatkaratekid committed:
make adjustments to introduction

fatkaratekid committed:
modify Technology breakdown

fatkaratekid committed:
modify execution engine

fatkaratekid committed:
modify integrated tools and final tech choice

fatkaratekid committed:
modify System Implementation and Demo Use Case

fatkaratekid committed:
modify conclusion

cbcunc commented:
@scoobies check references

scoobies commented:
Checking the BibTeX entries failed with the following error: 
```
Couldn't find bibliography entry
```

deniederhut commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,682 @@
+----------
+.. [bias] Steinbock, Bonnie (1978). *Speciesism and the Idea of Equality*, Philosophy, 53 (204): 247–256, doi:10.1017/S0031819100016582
+
+.. [aif360] AI Fairness 360 Toolkit, https://github.com/Trusted-AI/AIF360. Last accessed 18 Feb 2021
```
Please cite this software: https://github.com/Trusted-AI/AIF360#citing-aif360

deniederhut commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,682 @@
+
+.. [aif360] AI Fairness 360 Toolkit, https://github.com/Trusted-AI/AIF360. Last accessed 18 Feb 2021
+
+.. [aix360] AI Explainability 360 Toolkit, https://github.com/Trusted-AI/AIX360 Last accessed 18 Feb 2021
```
Please cite this software: https://github.com/Trusted-AI/AIX360#citing-ai-explainability-360

fatkaratekid replied:
fix it, thanks

deniederhut commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,682 @@
+
+.. [kubernetes] Kubernetes, https://kubernetes.io/. Last accessed 18 Feb 2021
+
+.. [jupyter] JupyterLab, https://jupyter.org/. Last accessed 18 Feb 2021
```
Please cite this software: https://github.com/jupyter/jupyter/issues/190#issuecomment-721264013

fatkaratekid replied:
For Jupyter they recommend this Zenodo DOI: 10.5281/zenodo.4539666 at this link https://jupyterbook.org/cite.html. Maybe we should go for this one, it respects the contributors' wishes. It is not aesthetically nice as your suggestion though.

fatkaratekid replied:
thanks for the ref

deniederhut commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,682 @@
+
+.. [kubeflow] https://www.kubeflow.org/ Last accessed 18 Feb 2021
+
+.. [katib] Katib, https://github.com/kubeflow/katib. Last accessed 18 Feb 2021
```
Please cite this software: https://github.com/kubeflow/katib#citation

fatkaratekid replied:
done 7375746
Thanks a lot for the link!

deniederhut reviewed (changes requested):
At least four of the entries in the bibliography are citing a website instead of the appropriate paper. Please look through your references list and check the rest of the entries.

fatkaratekid committed:
fix katib reference

fatkaratekid commented:
@deniederhut work on all the mentioned issues. For some of these it is really hard (I dare to say impossible) to find a publication with a DOI or at least arXiv. For some there is no excuse obviously. We'll give our best for sure. Thanks for pointing it out.

fatkaratekid committed:
fix Jupyter reference

fatkaratekid committed:
fix aix360 and jupyter reference

fatkaratekid committed:
fix references

fatkaratekid committed:
fix galaxy reference

fatkaratekid committed:
fix reana reference

fatkaratekid committed:
fix snakemake reference

Commit:
add art reference

Commit:
add reference to kubernetes

Commit:
add reference to kfserving

Commit:
add reference to kubeflow

Commit:
add reference to slurm

Commit:
add reference to snakemake

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,683 @@
+   supporting various computer languages, working on various data flow editors and
+   running on diverse execution engines.
+   To demonstrate its utility, we constructed a workflow composed exclusively of CLAIMED components.
+   For this purpose, we made use of a publicly available CT scans dataset [covidata]_
```
Computed Tomography (CT) scans... ?

romeokienzler replied:
thanks, fixed in 1f4d6f9459c2a7a4a2ba955ae34bacf28ba29323

Commit:
add reference to htcondor

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,683 @@
+Slurm [slurm]_, Snakemake [snakemake]_, QSub [qsub]_, HTCondor [htcondor]_, Apache Nifi [nifi]_,
+NodeRED [nodered]_, KNIME [knime]_, Galaxy [galaxy]_, 
+Reana [reana]_, WEKA [weka]_, Rabix [rabix]_, Nextflow [nextflow]_, OpenWDL [openwdl]_, CWL [cwl]_
+or Cromwell [cromwell]_.
```
This is an impressive list of tools, but I'm missing the information about why they are needed.
The answer is indeed in the next paragraph.
Perhaps it could be of help to clarify the production requirements first, then list the number of available tools. This will show how complex the picture is because of the requirements and how useful is a tool like the one you propose.

romeokienzler replied:
Thanks a lot for the suggestion. We've addressed your comment in 34778253a0fac2000650dd4bee1f8823e6a7f89d

Commit:
add reference to nodered

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,683 @@
+Before we address how CLAIMED fulfills the previously defined requirements and how the
+exemplary workflow has been constructed, we will introduce some terms and technologies.
+
+Technology breakdown
```
I like the choice for the title of this section.
My understanding is that a list of critical concepts for the implementation will follow: containerization, deep learning, ..., explainability, fairness.
Do you mind making more explicit what types of technology breakdown are covered by Kubeflow, Elyra, JupyterLab? In particular, is JupyterLab covering a technology breakdown or its place is more in the implementation part, among the requirements? As listed here
https://github.com/scipy-conference/scipy_proceedings/blob/e03cec9ff666247c8dff59bac975358198ea567d/papers/ivan_nesic/claimed.rst#L304-L316

romeokienzler replied:
Thanks a lot, we addressed your concerns in c39b989e070c5ca8d1e5dc35c7e7aa2a418d534e - please let us know it it fits to what you had in mind

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,683 @@
+- so called "Cloud Paks" - on top of Kubernetes, making them run everywhere.
+Therefore, IBM joined CNCF [ibmcncf]_, moved all Watson Services to Kubernetes
+and acquired RedHat. This makes IBM the 3rd largest committer to
+Kubernetes.
```
May I suggest to short sentences for readability?
Something like (I'm using github suggestions to make the diff explicit, not to feed you with my words... :) )
```suggestion
IBM acts as a Kubernetes runtime provider in the cloud and - 
through the acquisition of RedHat - has become the major vendor for 
on-premises Kubernetes. IBM is now able to deliver software solutions
- so called "Cloud Paks" - on top of Kubernetes, making them run everywhere.
Therefore, IBM joined CNCF [ibmcncf]_, and moved all Watson Services to Kubernetes.
This makes IBM the 3rd largest committer to Kubernetes.
```

romeokienzler replied:
Thanks, we've shortened in 9fbd4aee47d5ee0b5aae58376bbf659d7cef0aa5 to 

IBM acts as a Kubernetes runtime provider in the cloud and - 
through the acquisition of RedHat - has become the major vendor for 
on-premises Kubernetes. IBM is now able to deliver software solutions - so 
called "Cloud Paks" - on top of Kubernetes, making them run everywhere **(hybrid cloud)**.
Therefore, IBM joined CNCF [ibmcncf]_, and moved all Watson Services to Kubernetes.
This makes IBM the 3rd largest committer to Kubernetes.
**Not only for IBM but also for us, Kubernetes enables the hybrid cloud scenario of 
transparently moving workload
across different on-premises, remote and cloud data centers seamlessly.**

Just added "hybrid cloud" to your suggestion and a sentence at the bottom to close the loop of requirements to our use case as well.

cmarmo reviewed (commented):
Thanks a lot to the authors for your work on the proceeding: I find it clearer and easier to read, now.
I still have some comments about its first part, that I have detailed below.
I have not found the time to recheck the use case and the conclusions. I will do it ASAP. Sorry for that.

Commit:
add reference to knime

Commit:
add reference to rabix

Commit:
add reference to nextflow

Commit:
add reference to tensorflow eager

Commit:
add reference to tensorflow dataset

Commit:
add reference to tensorflow book

Commit:
expand CT scan

Commit:
revise introduction

fatkaratekid committed:
merge differences

fatkaratekid committed:
Merge branch '2021' of github.com:fatkaratekid/scipy_proceedings into 2021

deniederhut commented:
> For some of these it is really hard (I dare to say impossible) to find a publication with a DOI or at least arXiv.

Yeah, totally -- not everyone writes publications about their software, and even recent conference papers don't always get DOIs. You won't be able to find a good citation for everything, but we want to make our best effort, because it's important to recognize and celebrate the good work that others do.

Commit:
shorten passage on IBM's involvement in Kubernetes

Commit:
revise technology breakdown

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,703 @@
+
+Sometimes, transformations on the metadata, or any other structured
+dataset, are necessary. Therefore, we provide a generic transformation
+component - in the example, we just used it to change to format of the
```
```suggestion
component - in the example, we used it to change to format of the
```

romeokienzler replied:
thanks! fixed in 4c88ad49c34d4fcf39b600e31add50d8e2d56c91

Commit:
add figure to future work section

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,703 @@
+component - in the example, we just used it to change to format of the
+categories as the original file contained forward slashes which made it
+hard to use on the underlying operating system. This is performed by
+simply specifying a column name and a function that is to be applied.
```
```suggestion
specifying a column name and a function that is to be applied.
```

romeokienzler replied:
thanks! fixed in 4c88ad49c34d4fcf39b600e31add50d8e2d56c91

cmarmo commented on papers/ivan_nesic/claimed.rst:
```diff
@@ -0,0 +1,715 @@
+The name of the tool is C3 [c3]_, and it stands for CLAIMED component compiler.
+Currently, transpiling from notebooks to Kubeflow Pipeline components is supported.
+In addition, publishing these components to component repositories will also be possible.
+C3 already suports publishing components to Machine Learning Exchange (MLX) [mlx]_, an open source
```
```suggestion
C3 already supports publishing components to Machine Learning Exchange (MLX) [mlx]_, an open source
```

romeokienzler replied:
sorry for that! fixed in 4c88ad49c34d4fcf39b600e31add50d8e2d56c91

cmarmo reviewed (approved):
@fatkaratekid @romeokienzler thanks a lot for your work.
Two nitpicks and one typo, are my last comments.
Looking forward to see your contribution published.

Commit:
fix typos, improve language

Commit:
fix minor errors

fatkaratekid commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.3233/978-1-61499-649-1-87 is OK
- 10.1093/nar/gky379 is OK
- 10.1051/epjconf/201921406034 is OK
- 10.1093/bioinformatics/bts480 is OK
- 10.1109/ITHET.2017.8067827 is OK
- 10.1109/PDSW-DISCS.2018.00011 is OK
- 10.1109/SEAA51224.2020.00036 is OK

MISSING DOIs

- 10.1007/10968987_3 may be a valid DOI for title: SLURM: Simple Linux Utility for Resource Management
- 10.1007/978-3-030-19504-5_13 may be a valid DOI for title: Citizen Data Scientist: A Design Science Research Method for the Conduct of Data Science Projects

INVALID DOIs

- None
```

fatkaratekid commented:
I checked the 'missing' two and they are valid.

deniederhut commented:
Awesome! @stsievert do you feel this paper is now ready for inclusion in the proceedings?

fatkaratekid committed:
fix references format

deniederhut commented:
@scoobies mark ready

fatkaratekid commented:
@deniederhut Is it still possible to make minor modifications (fix typos and some reference formatting for instance )? If yes, I suppose this is to be done with a PR?
